File: ForthCPU/test/src/uart_echo.f
hex FFFF constant PORT
hex CFFF constant UART
decimal 128 constant MAX_COMMAND_LENGTH

0 ivariable led_state
0 ivariable timer_interrupt
0 ivariable command_ready
variable command_p
variable command_read_p
variable command_end
array command MAX_COMMAND_LENGTH

: isr1 1 timer_interrupt ! ;

: isr2
  UART @
  command_ready @ if0
    command_p @ command_end @ < if
      dup
      '\r' = if
        command_ready !
      else
        command_p @ ! command_p @ 1 + command_p !
      then
    then
  else
    drop
  then
;

: blink led_state @ dup PORT ! 1 + led_state ! ;

: uart_out begin PORT @ until UART ! ;

: uart_echo begin command_read_p @ command_p @ != while
    command_read_p @ dup @ uart_out
    1 + command_read_p !
  repeat
;

: main 
  command command_p !
  command command_read_p !
  command MAX_COMMAND_LENGTH + command_end !
  begin
    wfi
    timer_interrupt @ if
      0 timer_interrupt !
      blink
      uart_echo

      command_ready @ if
        '\r' uart_out '\n' uart_out
        command command_p !
        command command_read_p !
        0 command_ready !
      then
    then
  again
;

File: Software/JavaToForth/src/test/resources/UartEcho.java
class UartEcho {
    private static final int DELAY = 500 * Hal.MS;
    private static final int UART_BUFFER_SIZE = 128;

    private static boolean commandReady;
    private static char[] command;
    private static int commandPointer, commandReadPointer;

    private static boolean timerInterrupt;
    
    public static void isr1() {
        timerInterrupt = true;
    }

    public static void isr2() {
        char c = (char)(Hal.uartGet() & 0xFF);
        if (commandReady)
            return;
        if (c == '\n') {
            commandReady = true;
            return;
        }
        if (commandPointer < command.length) {
            command[commandPointer++] = c;
        }
    }

    private static void uartEcho() {
        while (commandReadPointer < commandPointer)
            Hal.outChar(command[commandReadPointer++]);
    }

    public static void main() {
        command = new char[UART_BUFFER_SIZE];
        commandReady = false;
        timerInterrupt = false;
        commandPointer = commandReadPointer = 0;
        boolean ledState = false;
        Hal.timerStart(DELAY);
        while (true) {
            System.wfi();
            uartEcho();
            if (timerInterrupt) {
                timerInterrupt = false;
                Hal.timerStart(DELAY);
                Hal.ledSet(ledState);
                ledState = !ledState;
                if (commandReady) {
                    commandReady = false;
                }
            }
        }
    }
}    
